proxy/proxy.go

```diff
@@ -462,177 +349,212 @@ func (p *Proxy) lookupRoute(r *http.Request) (rt *routing.Route, params map[stri
 	return p.routing.Route(r)
 }
 
-// send a premature error response
-func sendError(w http.ResponseWriter, error string, code int) {
-	http.Error(w, error, code)
+func sendHTTPError(w http.ResponseWriter, code int) {
 	addBranding(w.Header())
+	http.Error(w, http.StatusText(code), code)
 }
 
-// http.Handler implementation
-func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	startServe := time.Now()
+// send a premature error response
+func (p *Proxy) sendError(c *context, code int) {
+	sendHTTPError(c.responseWriter, code)
 
-	start := startServe
-	rt, params := p.lookupRoute(r)
-	if rt == nil {
-		if p.flags.Debug() {
-			dbgResponse(w, &debugInfo{
-				incoming: r,
-				response: &http.Response{StatusCode: http.StatusNotFound}})
-			return
-		}
+	id := unknownRouteId
+	if c.route != nil {
+		id = c.route.Id
+	}
 
-		p.metrics.IncRoutingFailures()
-		sendError(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
-		p.metrics.MeasureServe(unknownRouteId, r.Host, r.Method, http.StatusNotFound, startServe)
-		log.Debugf("Could not find a route for %v", r.URL)
+	p.metrics.MeasureServe(
+		id,
+		c.request.Host,
+		c.request.Method,
+		code,
+		c.startServe,
+	)
+}
+
+func (p *Proxy) makeUpgradeRequest(ctx *context, route *routing.Route, req *http.Request) {
+	// have to parse url again, because path is not copied by mapRequest
+	backendURL, err := url.Parse(route.Backend)
+	if err != nil {
+		log.Errorf("can not parse backend %s, caused by: %s", route.Backend, err)
+		sendHTTPError(ctx.responseWriter, http.StatusBadGateway)
 		return
 	}
 
-	p.metrics.MeasureRouteLookup(start)
+	reverseProxy := httputil.NewSingleHostReverseProxy(backendURL)
+	reverseProxy.FlushInterval = p.flushInterval
+	upgradeProxy := upgradeProxy{
+		backendAddr:     backendURL,
+		reverseProxy:    reverseProxy,
+		insecure:        p.flags.Insecure(),
+		tlsClientConfig: p.roundTripper.TLSClientConfig,
+	}
 
-	start = time.Now()
-	routeFilters := rt.Filters
-	c := p.newFilterContext(w, r, params, rt)
+	upgradeProxy.serveHTTP(ctx.responseWriter, req)
+	log.Debugf("finished upgraded protocol %s session", getUpgradeRequest(ctx.request))
 
-	var (
-		onErr        func(err interface{})
-		filterPanics []interface{}
-	)
-	if p.flags.Debug() {
-		onErr = func(err interface{}) {
-			filterPanics = append(filterPanics, err)
-		}
-	} else {
-		onErr = func(err interface{}) {
-			log.Error("filter", err)
+}
+
+func (p *Proxy) makeBackendRequest(ctx *context) (*http.Response, error) {
+	req, err := mapRequest(ctx.request, ctx.route, ctx.outgoingHost)
+	if err != nil {
+		log.Errorf("could not map backend request, caused by: %v", err)
+		p.sendError(ctx, http.StatusInternalServerError)
+		return nil, errProxyCanceled
+	}
+
+	if p.experimentalUpgrade && isUpgradeRequest(req) {
+		p.makeUpgradeRequest(ctx, ctx.route, req)
+		// We are not owner of the connection anymore.
+		return nil, errProxyCanceled
```

Comment:
errProxyCanceled means something like this: "It indicates to the caller that the current proxy mechanism is canceled due to varying reasons. If there were any errors, they are already handled."

Do you think it is OK to just put this comment next to the errProxyCanceled definition? Or can you suggest a better name?

Regarding a separate error object for the upgrade, it sounds legit, but it would come with the inconvenience that an additional case would need to be checked in the calling code.